marquez/airflow.py

```diff
@@ -1,93 +1,84 @@
+    _jobIdMapping = None
+    _mqz_client = None
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.mqz_client = MarquezClient()
         self.mqz_namespace = kwargs['default_args'].get('mqz_namespace', 'unknown')
         self.mqz_location = kwargs['default_args'].get('mqz_location', 'unknown')
         self.mqz_input_datasets = kwargs['default_args'].get('mqz_input_datasets', [])
         self.mqz_output_datasets = kwargs['default_args'].get('mqz_output_datasets', [])
+        self._jobIdMapping = JobIdMapping()
 
     def create_dagrun(self, *args, **kwargs):
-        job_name = self.dag_id
-        job_run_args = "{}"                     # TODO retrieve from DAG/tasks
-        start_time = pendulum.instance(kwargs['execution_date']).to_datetime_string()
-        end_time = None
-
-        self.mqz_client.set_namespace(self.mqz_namespace)
-        self.mqz_client.create_job(job_name, self.mqz_location, self.mqz_input_datasets, self.mqz_output_datasets,
-                                   self.description)
-        mqz_job_run_id = self.mqz_client.create_job_run(job_name, job_run_args=job_run_args,
-                                                        nominal_start_time=start_time,
-                                                        nominal_end_time=end_time).run_id
-        self.mqz_client.mark_job_run_running(mqz_job_run_id)
-
-        self.marquez_log('job_running', json.dumps(
-            {'namespace': self.mqz_namespace,
-             'name': job_name,
-             'description': self.description,
-             'location': self.mqz_location,
-             'runArgs': job_run_args,
-             'nominal_start_time': start_time,
-             'nominal_end_time': end_time,
-             'jobrun_id': mqz_job_run_id,
-             'inputDatasetUrns': self.mqz_input_datasets,
-             'outputDatasetUrns': self.mqz_output_datasets
-             }))
-
-        run = super().create_dagrun(*args, **kwargs)
-        airflow.models.Variable.set(run.run_id, mqz_job_run_id)
-
+        run_args = "{}"  # TODO extract the run Args from the tasks
+        mqz_job_run_id = self.report_jobrun(run_args, kwargs['execution_date'])
+        run = super(MarquezDag, self).create_dagrun(*args, **kwargs)
+        self._jobIdMapping.set(JobIdMapping.make_key(run.dag_id, run.run_id), mqz_job_run_id)
         return run
 
     def handle_callback(self, *args, **kwargs):
+        self.report_jobrun_change(args[0], **kwargs)
+        return super().handle_callback(*args, **kwargs)
+
+    def report_jobrun(self, run_args, execution_date):
         job_name = self.dag_id
-        mqz_job_run_id = self.get_and_delete(args[0].run_id)
+        job_run_args = run_args
+        start_time = pendulum.instance(execution_date).to_datetime_string()
+        end_time = pendulum.instance(self.following_schedule(execution_date)).to_datetime_string()
+        self.get_mqz_client().set_namespace(self.mqz_namespace)
```

Comment:
`self.get_mqz_client()` is called a few times. Why not just assign it to a variable and use that?